app/src/main/java/com/yalta/activities/fragments/MapFragment.kt

```diff
@@ -1,19 +1,136 @@
 package com.yalta.activities.fragments
 
+import android.Manifest
+import android.content.pm.PackageManager
+import android.location.Location
 import android.os.Bundle
 import androidx.fragment.app.Fragment
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import androidx.databinding.DataBindingUtil
+import androidx.lifecycle.ViewModelProvider
+import com.google.android.gms.maps.*
+import com.google.android.gms.maps.model.LatLng
 
 import com.yalta.R
+import com.yalta.databinding.FragmentMapBinding
+import com.yalta.viewmodel.MapViewModel
+
+class MapFragment : Fragment(), OnMapReadyCallback {
+    private val viewModel by lazy { ViewModelProvider(this).get(MapViewModel::class.java) }
+    private lateinit var mapView: MapView
+
+    private var location: Location? = null
+    private val defaultZoom = 16.0F
+    private var map: GoogleMap? = null
+    private var locationPermissionGranted = false
 
-class MapFragment : Fragment() {
 
     override fun onCreateView(
         inflater: LayoutInflater, container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
-        return inflater.inflate(R.layout.fragment_map, container, false)
+        val view = inflater.inflate(R.layout.fragment_map, container, false)
+
+        val binding: FragmentMapBinding =
+            DataBindingUtil.inflate(inflater, R.layout.fragment_map, container, false)
+
+        binding.lifecycleOwner = this
+        binding.viewModel = viewModel
+        binding.viewModel?.currentLocation?.observe(viewLifecycleOwner, { location ->
+            this.location = location
+            mapView.getMapAsync(this)
+        })
+
+        mapView = view.findViewById(R.id.map)
+        mapView.onCreate(savedInstanceState)
+        mapView.onResume()
+
+        mapView.getMapAsync(this)
+
+        return view
+    }
+
+    override fun onResume() {
+        super.onResume()
+        mapView.onResume()
+    }
+
+    override fun onPause() {
+        mapView.onPause()
+        super.onPause()
+    }
+
+    override fun onDestroy() {
+        mapView.onDestroy()
+        super.onDestroy()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        mapView.onSaveInstanceState(outState)
+    }
+
+    override fun onLowMemory() {
+        mapView.onLowMemory()
+        super.onLowMemory()
+    }
+
+    override fun onMapReady(googleMap: GoogleMap) {
+        map = googleMap
+        getLocationPermission()
```

Comment:
Looks like bad design. It looks like we call a function, it produces a side-effect and we have to know what exactly it does to check it. 
Maybe `getLocationPermission()` should return? (as well as set internal flag?) (Or it should only be called when flag property is accessed?)

Reply:
Actually... can we control order of execution here? is it guaranteed that it will be:
```
map = googleMap
---
getLocationPermission()
---
onRequestPermissionsResult
---
if (locationPermissionGranted) { 
```
Otherwise we'll always check trash after immediate return from `getLocationPermission`

Reply:
> Maybe `getLocationPermission()` should return? (as well as set internal flag?) (Or it should only be called when flag property is accessed?)
Ok, I get that we can't return from there...